consensus/istanbul/backend/announce.go

```diff
@@ -17,208 +17,386 @@
+					currentAnnounceGossipTickerDuration = time.Duration(sb.config.AnnounceGossipPeriod) * time.Second
+				}
+
+				// Enable periodic gossiping by setting announceGossipTickerCh to non nil value
+				announceGossipTicker = time.NewTicker(currentAnnounceGossipTickerDuration)
+				announceGossipTickerCh = announceGossipTicker.C
+				logger.Trace("Enabled periodic gossiping of announce message")
+			} else if !shouldAnnounce && announceGossipTickerCh != nil {
+				// Disable periodic gossiping by setting announceGossipTickerCh to nil
+				announceGossipTicker.Stop()
+				announceGossipTickerCh = nil
+				logger.Trace("Disabled periodic gossiping of announce message")
+			}
+
+		case <-announceGossipTickerCh: // If this is nil (when shouldAnnounce was most recently false), this channel will never receive an event
+			logger.Trace("Going to gossip an announce message", "announceGossipFrequencyState", announceGossipFrequencyState, "numGossipedMsgsInHighFreqAfterFirstPeerState", numGossipedMsgsInHighFreqAfterFirstPeerState)
+			switch announceGossipFrequencyState {
+			case HighFreqBeforeFirstPeerState:
+				if len(sb.broadcaster.FindPeers(nil, p2p.AnyPurpose)) > 0 {
+					announceGossipFrequencyState = HighFreqAfterFirstPeerState
+				}
+
+			case HighFreqAfterFirstPeerState:
+				if numGossipedMsgsInHighFreqAfterFirstPeerState >= 10 {
+					announceGossipFrequencyState = LowFreqState
+				}
+				numGossipedMsgsInHighFreqAfterFirstPeerState += 1
+
+			case LowFreqState:
+				if currentAnnounceGossipTickerDuration != time.Duration(sb.config.AnnounceGossipPeriod)*time.Second {
+					// Reset the ticker
+					currentAnnounceGossipTickerDuration = time.Duration(sb.config.AnnounceGossipPeriod) * time.Second
+					announceGossipTicker.Stop()
+					announceGossipTicker = time.NewTicker(currentAnnounceGossipTickerDuration)
+					announceGossipTickerCh = announceGossipTicker.C
+				}
+			}
+
+			go sb.generateAndGossipAnnounce()
+
+			// Use this timer to also prune all announce related data structures.
+			if err := sb.pruneAnnounceDataStructures(); err != nil {
+				logger.Warn("Error is pruning announce data structures", "err", err)
+			}
+
+		case <-announceVersionsCheckTicker.C:
+			logger.Trace("Going to check peers' announce version set")
+			go sb.checkPeersAnnounceVersions()
+
+		case <-sb.announceThreadQuit:
+			checkIfShouldAnnounceTicker.Stop()
+			announceVersionsCheckTicker.Stop()
+			if announceGossipTicker != nil {
+				announceGossipTicker.Stop()
+			}
+			return
+		}
+	}
 }
 
-type announceData struct {
-	AnnounceRecords []*announceRecord
-	EnodeURLHash    common.Hash
-	Timestamp       uint
+func (sb *Backend) shouldGenerateAndProcessAnnounce() (bool, error) {
+
+	// Check if this node is in the registered/elected validator set
+	regAndElectedVals, err := sb.retrieveRegisteredAndElectedValidators()
+	if err != nil {
+		return false, err
+	}
+
+	return sb.coreStarted && regAndElectedVals[sb.Address()], nil
```

Comment:
is the core mutex needed here?

Reply:
I actually think it's not, because the thread that uses it runs this once every 5 minutes.  If it's "off" during one of those runs, then the thread will correct it's understanding very quickly.  And it's not strictly necessary that this variable is consistent w/ the announce protocol.  Worst case is that a node will send out an extra announce message when it's not needed.